### DISCUSSION TITLE
Limitation through database specific migration system

### DISCUSSION THREAD

**Discussion by luwol03** [prisma/prisma#12568] [Q&A]
Hello, I'm currently playing the first time with prisma. I like the schema file and all those ts types that are generated. But one thing I dont understand, how can I use different database dialects. Because with ORM's I used previous, I was able to develop with an sqlite db without setting up a psql database. But the migration system from prisma limits that great ability prisma has to one db dialect at the time. Because if I understand that correctly, I have to remove all migration files and regenerate for a new dialect.
(For me its like an not cheap all round car that can drive on several different undergrounds (water, slum, street, hills, desert, ...), but currently it is limited to one underground, because I everytime have to change the wheels to be able to drive on another underground.)

I jumped from sequelize (with umzug as a migration system) to prisma, because of that typesafty and gql integration. Sequelize uses a query builder approach, so that it is possible to define the migrations with that query builder and run that on different dialects with umzug while migration.

**Comment by austincrim** [comment]
Hey @luwol03 👋 ,

If you want to use SQLite during development and Postgres in production, you could use the [`prisma db push`](https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push) command to make schema changes in SQLite without generating migration files. `db push` and [integrated seeding](https://www.prisma.io/docs/guides/database/seed-database) were built specifically for better prototyping and development workflows.

Does that fit your use case?

**Reply by luwol03** [reply]
Hey, thank you for your quick replay,

Sorry to say, but not really. I want to be also able to use sqlite for some small setups of my application in production. E.g. on Windows as a simple Desktop App in background (with electron, but also on real servers with psql)

But can you explain the reason, why the migration system limits that great ability of an orm. Why has Prisma decided for that approach?

**Reply by austincrim** [reply]
Ah, I see, so you're interested in actually using multiple different databases in production with the same Prisma schema.

This is not supported yet but has been on our radar for awhile. Feel free to upvote [this issue](https://github.com/prisma/prisma/issues/2443) and leave your use case if you feel it hasn't been covered yet in the discussion. Our engineering teams lean on those issues to prioritize and design upcoming features!

**Reply by luwol03** [reply]
Oh, so it's not outside of Prismas project scope, but not implemented currently?

But your linked issue looks more like having different datasources at the same time? I'm thinking more like having only one datasource that is able to connect to different database systems like it is already possible without using migrations. So I think the blocking factor is the migration system as I said (the car and the wheels). 
I currently think about two solutions, but are they in Prismas project scope/roadmap?
1.having different migration sql files side by side, for all supported dialects
2. having a query builder or some migration instructions sheet format

**Reply by janpio** [reply]
Prisma used to support this use case in the past, but does not any more. 

The Prisma schema can also be significantly different between databases (e.g. native types). If you make sure your schema is identical for all of them, and you find your own way to migrate them (or maybe store the migrations in different branches of the projects etc), you can probably get to what you want in some hacky way - but it is not what we suggest.

**Reply by luwol03** [reply]
What were Prismas thoughts to remove that functionality?

**Reply by janpio** [reply]
Very short version: It gave developers a false sense of security that testing with SQLite would be a good idea if you then deploy to PostgreSQL and run that in production. It usually is not.

**Reply by luwol03** [reply]
But why hardly limiting that, instead of letting the user decide what he likes to do. I can easily run my tests with a matrix on all different database systems.

**Reply by janpio** [reply]
Because it of course also limited what we can do, and how we can do it. Our current Migration tool for example can only create SQL for 1 SQL dialect as it needs to persist the files on disk. (That was different 2 years ago in the previous iteration of a Migration tool)

**Reply by luwol03** [reply]
What do you suggest to solve my problem? How can I run my little Server locally in an electron app with sqlite and on a server with Postgres? Wouldn't it be a good option to be able to create migration files for different dialects in different folders?

**Reply by janpio** [reply]
Yes, that would be optimal for your use case - but it is not one of the currently supported use cases as it is seldomly (almost never) asked for. You would have to find a way to hack that together for now, which probably should be possible with a small script that adapts the schema and moves the migration files around.

You might also want to open a feature request - your use case of wanting to use the same Prisma schema on both a client (via Electron with SQLite) and server might become more and more important and common. A feature request is a good way to track the demand.

**Reply by luwol03** [reply]
Wouldn't it be as easy as just add a cli flag for the migrations path and then change this line?
https://github.com/prisma/prisma/blob/8c94387faa063399aca8b920f2148226ef0023cc/packages/migrate/src/Migrate.ts#L25

**Reply by janpio** [reply]
Check if there is a feature requests for this, and if not please open one. (And no, a change is never that simple in the end 😆)

**Reply by luwol03** [reply]
There is one, I found it. It is from you. https://github.com/prisma/prisma/issues/4720

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
